{
    "active_lock_reason": null,
    "assignee": null,
    "assignees": [],
    "author_association": "NONE",
    "body": "When defining a cpdef enum in Cython, the enums seems to be considered as an int, meaning they are considered to be 32bit fields. Which is not always correct, for example if we have an external header file (test.h) with the following enum definition:\r\n```C\r\nenum {\r\n        ENUM_1 = ((uint64_t)(1 << 63)),\r\n};\r\n```\r\n\r\nAnd if I try in my cython_test.pxd/pyx files to:\r\n```Cython\r\ncdef extern from 'test.h':\r\n        cpdef enum:\r\n                 ENUM_1\r\n```\r\n\r\nThe compilation ends with the warning:\r\n```\r\nIn function 'PyInit_cython_test':\r\n warning: overflow in implicit constant conversion [-Woverflow]\r\nPyObject* wrapped = __Pyx_PyInt_From_int(ENUM_1);\r\n```\r\n\r\nIf I print the value on ENUM_1 in Cython's code it's really a 32bit, and the value of the original enum is cut !",
    "closed_at": null,
    "comment_data": [
        {
            "author_association": "NONE",
            "body": "You are not complying with the C standard. `enum`s should not be 64 bits.\r\nhttps://stackoverflow.com/questions/366017/what-is-the-size-of-an-enum-in-c",
            "created_at": "2019-11-20T05:00:54Z",
            "html_url": "https://github.com/cython/cython/issues/3240#issuecomment-555838651",
            "id": 555838651,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/3240",
            "node_id": "MDEyOklzc3VlQ29tbWVudDU1NTgzODY1MQ==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/cython/cython/issues/comments/555838651/reactions"
            },
            "updated_at": "2019-11-20T05:00:54Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/555838651",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/5534781?v=4",
                "events_url": "https://api.github.com/users/leofang/events{/privacy}",
                "followers_url": "https://api.github.com/users/leofang/followers",
                "following_url": "https://api.github.com/users/leofang/following{/other_user}",
                "gists_url": "https://api.github.com/users/leofang/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/leofang",
                "id": 5534781,
                "login": "leofang",
                "node_id": "MDQ6VXNlcjU1MzQ3ODE=",
                "organizations_url": "https://api.github.com/users/leofang/orgs",
                "received_events_url": "https://api.github.com/users/leofang/received_events",
                "repos_url": "https://api.github.com/users/leofang/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/leofang/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/leofang/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/leofang"
            }
        },
        {
            "author_association": "NONE",
            "body": "There are compiler extensions where larger (than int) enum values can be supported.\r\nBut Cython seems not to support that. ",
            "created_at": "2019-11-21T16:39:28Z",
            "html_url": "https://github.com/cython/cython/issues/3240#issuecomment-557168768",
            "id": 557168768,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/3240",
            "node_id": "MDEyOklzc3VlQ29tbWVudDU1NzE2ODc2OA==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/cython/cython/issues/comments/557168768/reactions"
            },
            "updated_at": "2019-11-21T16:39:28Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/557168768",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/44767618?v=4",
                "events_url": "https://api.github.com/users/EdwardSro/events{/privacy}",
                "followers_url": "https://api.github.com/users/EdwardSro/followers",
                "following_url": "https://api.github.com/users/EdwardSro/following{/other_user}",
                "gists_url": "https://api.github.com/users/EdwardSro/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/EdwardSro",
                "id": 44767618,
                "login": "EdwardSro",
                "node_id": "MDQ6VXNlcjQ0NzY3NjE4",
                "organizations_url": "https://api.github.com/users/EdwardSro/orgs",
                "received_events_url": "https://api.github.com/users/EdwardSro/received_events",
                "repos_url": "https://api.github.com/users/EdwardSro/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/EdwardSro/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/EdwardSro/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/EdwardSro"
            }
        },
        {
            "author_association": "NONE",
            "body": "``` C\r\nenum {\r\n    ONE = 1U << 31\r\n};\r\n```\r\n\r\n``` Cython\r\ncpdef enum:\r\n    ONE\r\n```\r\n\r\n``` Python\r\nprint(ONE)  # -2147483648\r\n```\r\nThis is a major problem. Cython needs to either allow some way to manually set `enum` type or allow large number to pass through without converting to negative value.\r\n\r\nEdit: Also I am not even getting any warning/error, letting me know there was an OverFlow in `cpdef enum`, it just converted silently. This leads to lots of bugs!\r\n\r\nEdit: using `Cython 3.0.8`",
            "created_at": "2024-02-20T19:30:32Z",
            "html_url": "https://github.com/cython/cython/issues/3240#issuecomment-1954917584",
            "id": 1954917584,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/3240",
            "node_id": "IC_kwDOABDGAc50hazQ",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/cython/cython/issues/comments/1954917584/reactions"
            },
            "updated_at": "2024-02-20T19:38:03Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/1954917584",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/35307184?v=4",
                "events_url": "https://api.github.com/users/YoSTEALTH/events{/privacy}",
                "followers_url": "https://api.github.com/users/YoSTEALTH/followers",
                "following_url": "https://api.github.com/users/YoSTEALTH/following{/other_user}",
                "gists_url": "https://api.github.com/users/YoSTEALTH/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/YoSTEALTH",
                "id": 35307184,
                "login": "YoSTEALTH",
                "node_id": "MDQ6VXNlcjM1MzA3MTg0",
                "organizations_url": "https://api.github.com/users/YoSTEALTH/orgs",
                "received_events_url": "https://api.github.com/users/YoSTEALTH/received_events",
                "repos_url": "https://api.github.com/users/YoSTEALTH/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/YoSTEALTH/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/YoSTEALTH/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/YoSTEALTH"
            }
        },
        {
            "author_association": "MEMBER",
            "body": "I think part of your problem is that you're using anonymous enums.\r\n\r\nWith non-anonymous enums you get a conversion function `__Pyx_PyInt_from_enum_type` which takes an object of the enum type in outputs a Python object. For compilers that support wider enums that has a decent chance of being right (I think).\r\n\r\nWith anonymous enums there just isn't a named type that the conversion function can take so I don't see what else Cython can do.",
            "created_at": "2024-02-20T22:16:04Z",
            "html_url": "https://github.com/cython/cython/issues/3240#issuecomment-1955210692",
            "id": 1955210692,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/3240",
            "node_id": "IC_kwDOABDGAc50iiXE",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/cython/cython/issues/comments/1955210692/reactions"
            },
            "updated_at": "2024-02-20T22:16:04Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/1955210692",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/10536947?v=4",
                "events_url": "https://api.github.com/users/da-woods/events{/privacy}",
                "followers_url": "https://api.github.com/users/da-woods/followers",
                "following_url": "https://api.github.com/users/da-woods/following{/other_user}",
                "gists_url": "https://api.github.com/users/da-woods/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/da-woods",
                "id": 10536947,
                "login": "da-woods",
                "node_id": "MDQ6VXNlcjEwNTM2OTQ3",
                "organizations_url": "https://api.github.com/users/da-woods/orgs",
                "received_events_url": "https://api.github.com/users/da-woods/received_events",
                "repos_url": "https://api.github.com/users/da-woods/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/da-woods/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/da-woods/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/da-woods"
            }
        },
        {
            "author_association": "NONE",
            "body": "@da-woods \r\n> I think part of your problem is that you're using anonymous enums.\r\n\r\nAww... you must be kidding me, I spent so much time trying to debug this issue! lol Thanks, that actually does fix the problem.\r\n\r\nWhat if you are using enum to define:\r\n``` Cython\r\ncdef extern from \"<limits.h>\" nogil:\r\n\r\n    cpdef enum:\r\n        UINT_MAX\r\n```\r\n``` Python\r\nprint(UINT_MAX)  # returns `-1` rather than `4294967295`\r\n```",
            "created_at": "2024-02-20T23:02:08Z",
            "html_url": "https://github.com/cython/cython/issues/3240#issuecomment-1955265135",
            "id": 1955265135,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/3240",
            "node_id": "IC_kwDOABDGAc50ivpv",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/cython/cython/issues/comments/1955265135/reactions"
            },
            "updated_at": "2024-02-20T23:02:08Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/1955265135",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/35307184?v=4",
                "events_url": "https://api.github.com/users/YoSTEALTH/events{/privacy}",
                "followers_url": "https://api.github.com/users/YoSTEALTH/followers",
                "following_url": "https://api.github.com/users/YoSTEALTH/following{/other_user}",
                "gists_url": "https://api.github.com/users/YoSTEALTH/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/YoSTEALTH",
                "id": 35307184,
                "login": "YoSTEALTH",
                "node_id": "MDQ6VXNlcjM1MzA3MTg0",
                "organizations_url": "https://api.github.com/users/YoSTEALTH/orgs",
                "received_events_url": "https://api.github.com/users/YoSTEALTH/received_events",
                "repos_url": "https://api.github.com/users/YoSTEALTH/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/YoSTEALTH/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/YoSTEALTH/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/YoSTEALTH"
            }
        }
    ],
    "comments": 5,
    "comments_url": "https://api.github.com/repos/cython/cython/issues/3240/comments",
    "created_at": "2019-11-19T14:56:42Z",
    "event_data": [
        {
            "actor": {
                "avatar_url": "https://avatars.githubusercontent.com/u/10536947?v=4",
                "events_url": "https://api.github.com/users/da-woods/events{/privacy}",
                "followers_url": "https://api.github.com/users/da-woods/followers",
                "following_url": "https://api.github.com/users/da-woods/following{/other_user}",
                "gists_url": "https://api.github.com/users/da-woods/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/da-woods",
                "id": 10536947,
                "login": "da-woods",
                "node_id": "MDQ6VXNlcjEwNTM2OTQ3",
                "organizations_url": "https://api.github.com/users/da-woods/orgs",
                "received_events_url": "https://api.github.com/users/da-woods/received_events",
                "repos_url": "https://api.github.com/users/da-woods/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/da-woods/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/da-woods/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/da-woods"
            },
            "commit_id": null,
            "commit_url": null,
            "created_at": "2024-02-20T23:02:10Z",
            "event": "mentioned",
            "id": 11868375526,
            "node_id": "MEE_lADOABDGAc4fS86-zwAAAALDaQnm",
            "performed_via_github_app": null,
            "url": "https://api.github.com/repos/cython/cython/issues/events/11868375526"
        },
        {
            "actor": {
                "avatar_url": "https://avatars.githubusercontent.com/u/10536947?v=4",
                "events_url": "https://api.github.com/users/da-woods/events{/privacy}",
                "followers_url": "https://api.github.com/users/da-woods/followers",
                "following_url": "https://api.github.com/users/da-woods/following{/other_user}",
                "gists_url": "https://api.github.com/users/da-woods/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/da-woods",
                "id": 10536947,
                "login": "da-woods",
                "node_id": "MDQ6VXNlcjEwNTM2OTQ3",
                "organizations_url": "https://api.github.com/users/da-woods/orgs",
                "received_events_url": "https://api.github.com/users/da-woods/received_events",
                "repos_url": "https://api.github.com/users/da-woods/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/da-woods/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/da-woods/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/da-woods"
            },
            "commit_id": null,
            "commit_url": null,
            "created_at": "2024-02-20T23:02:10Z",
            "event": "subscribed",
            "id": 11868375539,
            "node_id": "SE_lADOABDGAc4fS86-zwAAAALDaQnz",
            "performed_via_github_app": null,
            "url": "https://api.github.com/repos/cython/cython/issues/events/11868375539"
        }
    ],
    "events_url": "https://api.github.com/repos/cython/cython/issues/3240/events",
    "html_url": "https://github.com/cython/cython/issues/3240",
    "id": 525061822,
    "labels": [],
    "labels_url": "https://api.github.com/repos/cython/cython/issues/3240/labels{/name}",
    "locked": false,
    "milestone": null,
    "node_id": "MDU6SXNzdWU1MjUwNjE4MjI=",
    "number": 3240,
    "performed_via_github_app": null,
    "reactions": {
        "+1": 0,
        "-1": 0,
        "confused": 0,
        "eyes": 0,
        "heart": 0,
        "hooray": 0,
        "laugh": 0,
        "rocket": 0,
        "total_count": 0,
        "url": "https://api.github.com/repos/cython/cython/issues/3240/reactions"
    },
    "repository_url": "https://api.github.com/repos/cython/cython",
    "state": "open",
    "state_reason": null,
    "timeline_url": "https://api.github.com/repos/cython/cython/issues/3240/timeline",
    "title": "Wrong enum conversion with cpdef enum (Cython language level=3)",
    "updated_at": "2024-02-20T23:02:10Z",
    "url": "https://api.github.com/repos/cython/cython/issues/3240",
    "user": {
        "avatar_url": "https://avatars.githubusercontent.com/u/44767618?v=4",
        "events_url": "https://api.github.com/users/EdwardSro/events{/privacy}",
        "followers_url": "https://api.github.com/users/EdwardSro/followers",
        "following_url": "https://api.github.com/users/EdwardSro/following{/other_user}",
        "gists_url": "https://api.github.com/users/EdwardSro/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/EdwardSro",
        "id": 44767618,
        "login": "EdwardSro",
        "node_id": "MDQ6VXNlcjQ0NzY3NjE4",
        "organizations_url": "https://api.github.com/users/EdwardSro/orgs",
        "received_events_url": "https://api.github.com/users/EdwardSro/received_events",
        "repos_url": "https://api.github.com/users/EdwardSro/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/EdwardSro/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/EdwardSro/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/EdwardSro"
    }
}